src/library/scala/collection/immutable/HashSet.scala

```diff
@@ -90,9 +90,17 @@ sealed class HashSet[A] extends AbstractSet[A]
 
   override def union(that: GenSet[A]): HashSet[A] = that match {
     case that: HashSet[A] =>
-      val buffer = new Array[HashSet[A]](bufferSize(this.size + that.size))
-      nullToEmpty(union0(that, 0, buffer, 0))
-    case _ => super.union(that)
+      if (this eq that) this
+      else nullToEmpty(union0(that, 0))
+    case _ =>
+      if (that.isEmpty) this else {
+        object acc extends Function1[A, Unit] {
+          var res = HashSet.this
+          override def apply(v1: A): Unit = res += v1
+        }
+        that foreach acc
```

Comment:
`foreach` can run in parallel for a `ParSet`. This implementation should be restricted to `Set`.

Reply:
reworked as 
```
        if (that.isInstanceOf[Set[A]])
          that foreach acc
        else
          that.iterator foreach acc

```